### PR TITLE
Add custom events adding UI

### LINKED ISSUE #601: Enable reporting CTR on sidebar and navbar links
I'd be beneficial to a publisher to be able to see CTRs on links in navbar and sidebar. 

Example: a site displays a "Donate" link in the top navbar. A publisher would like to know how many people, out of all who seen it, have clicked the link. 

## Technical implementation

Basically it's about adding analytics events via [the `newspack_analytics_events` filter](https://github.com/Automattic/newspack-plugin/blob/9eaaf1842619110effca4a873b64ab6d85f67915/includes/class-analytics.php#L175) - an impression event, and a click event. 
The tricky part is choosing when to do it – sending these events for each link might be too much and spam the GA data. A simple solution would be to have a setting for that, off by default. A more granular solution would be to enable choosing exactly which links should have those events attached. 
We also might consider displaying the CTR data in Newspack UI.

### PR DESCRIPTION
**Pull request by adekbadek**
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Adding custom events insertion UI. Only `click` and `submit` events are handled: 
- `visibility`, `ini-load` require the element to be AMP element,
- `scroll` requires some more UI for scroll parameters, can be implemented later.

Design considerations:

1. This is an advanced feature, misuse of which can brake sites – we may wanna add some warning there
2. Some fields in the form are not required, not sure if we have an established pattern for that
3. The "Active" field is of a different kind, maybe it should be displayed differently in the form? 

Closes #601

### How to test the changes in this Pull Request:

1. Open the Analytics Wizard, observe a new tab – "Custom Events"
2. Add a custom event, e.g. a click on `#search-toggle`
3. Visit the front end (as non-logged-in user), oberve the event being sent on action (look for requests to `https://www.google-analytics.com/r/collect` in Network tab) 
4. Test on both AMP and non-AMP versions
4. Mark the event as not active, observe it's not sent anymore

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

File: includes/wizards/class-analytics-wizard.php [hunk 1]
```diff
@@ -311,4 +368,28 @@ public static function set_category_dimension( $request ) {
...
+					$event['id'] = uniqid();
+				}
+				return $event;
+			},
+			$request['events']
+		);
+		if ( update_option( self::$custom_events_option_name, wp_json_encode( $custom_events ) ) ) {
+			return [ 'events' => $custom_events ];
+		} else {
+			return new WP_Error( 'newspack_analytics', __( 'Error when setting category custom dimension.', 'newspack' ) );
```
[the remark this answers is not in the document]
Comment (adekbadek) [reply includes/wizards/class-analytics-wizard.php]: Good catch, fixed in 17a4498

**Comment by adekbadek** [comment]
> How badly can this feature break the site? Is it that your analytics would stop working, or can you actually break site functionality with this?

No, nothing bad can happen. It's more about the user not being surprised by this developer-like options. I think it will be enough to say "This is an advanced feature, read more about on our support page" – with docs link as you suggested:

image

> As an advanced section, it could also help if it's not the first/default tab. 

Good call, updated in 3bcb711a8af62e5cb4d40fb8d7faaba2b1b81e73

> I think we could keep it simple visually, but I'm not sure complicated doing something like this would be to do correctly (eg. have text, or a CSS class, be appended automatically based on the 'required' status):

Done in a pretty hacky way in 603f02145d83ddd57a780a72287a74af295efafd

> How is [active field] different from the other fields?

Should've been clearer about this, sorry :) It's different because "active" is not a property of a GA custom event, it's just for the use of the plugin. A user might want to have the custom event defined, but not reported temporarily – in such a case they can mark it as inactive, like a post draft. But it's not a big issue, just looked a bit odd in the modal:


> but the checkbox was disabled for me after creating new events

That's on me – the table row is clickable, and then you can click the checkbox in a modal.  I've changed it so there are "Edit" links in the table, and an icon instead of a disabled checkbox:

image

> Lastly, tiny nitpick, what do you think about changing the 'Non-interaction' label to 'Non-interaction event' (just to match Google's docs)?

Sure, updated in 3bcb711a8af62e5cb4d40fb8d7faaba2b1b81e73

**Comment by adekbadek** [comment]
> How badly can this feature break the site? Is it that your analytics would stop working, or can you actually break site functionality with this?

No, nothing bad can happen. It's more about the user not being surprised by this developer-like options. I think it will be enough to say "This is an advanced feature, read more about on our support page" – with docs link as you suggested:

image

Note that the docs page is a draft at the moment.

> As an advanced section, it could also help if it's not the first/default tab. 

Good call, updated in 90f67a5515fc2083baa7e4cdefdb863c1642acd7

> I think we could keep it simple visually, but I'm not sure complicated doing something like this would be to do correctly (eg. have text, or a CSS class, be appended automatically based on the 'required' status):

Done in a pretty hacky way in 0a0fd687c3a58de4cd4d965bc28e87f82dfb228e

> How is [active field] different from the other fields?

Should've been clearer about this, sorry :) It's different because "active" is not a property of a GA custom event, it's just for the use of the plugin. A user might want to have the custom event defined, but not reported temporarily – in such a case they can mark it as inactive, like a post draft. But on a second thought, I don't think it's a big issue.

> but the checkbox was disabled for me after creating new events

That's on me – the table row was clickable, and the checkbox was in the way.  I've changed it so there are "Edit" links in the table, and an icon instead of a disabled checkbox:

image

> Lastly, tiny nitpick, what do you think about changing the 'Non-interaction' label to 'Non-interaction event' (just to match Google's docs)?

Sure, updated in 90f67a5515fc2083baa7e4cdefdb863c1642acd7

File: assets/components/src/text-control/index.js [hunk 2]
```diff
@@ -19,13 +20,31 @@ import './style.scss';
...
 	 * Render.
 	 */
 	render() {
-		const { className, ...otherProps } = this.props;
+		const { className, required, ...otherProps } = this.props;
 		const classes = classNames( 'newspack-text-control', className );
-		return <BaseComponent className={ classes } { ...otherProps } />;
+		return required ? (
+			<div ref={ this.wrapperRef } className={ classes }>
+				<BaseComponent required={ required } { ...otherProps } />
```
Comment (laurelfulford) [inline assets/components/src/text-control/index.js:43]: The current placement of `className` on require text controls causes `.newspack-text-control` not to be added to `div.components-base-control`, which makes the labels the wrong font size -- they're `13px` rather `14px`:

image

Moving the `className` to the `BaseComponent` seems to fix, but I'm not sure if this is a bad idea for a different reason:

```suggestion
			<div ref={ this.wrapperRef }>
				<BaseComponent className={ classes } required={ required } { ...otherProps } />
```

File: assets/components/src/text-control/style.scss [hunk 3]
```diff
@@ -11,12 +11,22 @@
...
 		font-weight: bold;
 		line-height: inherit;
 		margin: 0;
+
+		&[data-required-text] {
+			&::after {
+				content: attr( data-required-text );
+				padding-left: 5px;
+				font-weight: normal;
+				opacity: 0.5;
```
Comment (laurelfulford) [inline assets/components/src/text-control/style.scss:27]: Minor nitpick, but suggest swapping the opacity here for one of the existing mid-greys used in the UI, like:

```suggestion
				 color: $dark-gray-300;
```

File: assets/wizards/analytics/views/custom-events/index.js [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+						'newspack'
+					) }
+				</p>
+				<Notice
+					rawHTML
+					isInfo
+					noticeText={ `${ __(
+						'This is an advanced feature, read more about it on our',
+						'newspack'
+					) } <a href="${ NEWSPACK_SUPPORT_URL }/analytics">${ __(
```
Comment (laurelfulford) [inline assets/wizards/analytics/views/custom-events/index.js:138]: Another really minor nitpick: 

```suggestion
					) } <a target="_blank" href="${ NEWSPACK_SUPPORT_URL }/analytics">${ __(
```

File: assets/components/src/text-control/index.js (hunk 2 above)
Comment (adekbadek) [reply assets/components/src/text-control/index.js]: Done in 5ae9711750d7e43a80fd15b32bcfed700c7e849b

File: assets/components/src/text-control/style.scss (hunk 3 above)
Comment (adekbadek) [reply assets/components/src/text-control/style.scss]: Done in 5ae9711750d7e43a80fd15b32bcfed700c7e849b

File: assets/wizards/analytics/views/custom-events/index.js (hunk 4 above)
Comment (adekbadek) [reply assets/wizards/analytics/views/custom-events/index.js]: Done in 5ae9711750d7e43a80fd15b32bcfed700c7e849b

**Comment by adekbadek** [comment]
@laurelfulford 

> It looks like those docs still need to be written

Yes, I've started a draft but will publish when these changes are released.

File: assets/components/src/text-control/index.js (hunk 2 above)
Comment (laurelfulford) [reply assets/components/src/text-control/index.js]: Looks good, thanks!

File: assets/components/src/text-control/style.scss (hunk 3 above)
Comment (laurelfulford) [reply assets/components/src/text-control/style.scss]: Thanks!

File: assets/wizards/analytics/views/custom-events/index.js (hunk 4 above)
Comment (laurelfulford) [reply assets/wizards/analytics/views/custom-events/index.js]: Thanks Adam!

**Pull request merged**